{ai}[gfbf/2024a] jax v0.4.35 w/ CUDA 12.6.0

ThomasHoffmann77 opened the pull request:
(created using `eb --new-pr`)

fizwit commented:
where is the easyconfig for Clang-18.1.8-gfbf-2024a-CUDA-12.6.0.eb

ThomasHoffmann77 commented:
> where is the easyconfig for Clang-18.1.8-gfbf-2024a-CUDA-12.6.0.eb

#22300

boegel commented on easybuild/easyconfigs/b/Bazel/Bazel-6.5.0-GCCcore-13.3.0.eb:
```diff
@@ -0,0 +1,31 @@
+]
+
+dependencies = [
+    ('Java', '11.0.20', '', SYSTEM),
```
Can we use Java/21 here?

ThomasHoffmann77 replied:
@boegel Build of Bazel 6.5.0 fails for both, Java/21 and Java/21.0.2.

ThomasHoffmann77 commented on easybuild/easyconfigs/j/jax/jax-0.4.34-gfbf-2024a-CUDA-12.6.0.eb:
```diff
@@ -0,0 +1,229 @@
+cuda_compute_capabilities = ["5.0", "6.0", "6.1", "7.0", "7.5", "8.0", "8.6", "9.0"]
+
+builddependencies = [
+    # ('Bazel', '7.4.1'),  TODO: problems with @@local_config_python//:py3_runtime:
```
@boegel: I am not an bazel expert, but I think it is rather a fundamental incompatibility as the latest jax 0.5.2 still is using Bazel v6.5.0: https://github.com/jax-ml/jax/blob/ce224293b1a7d9b39b5d9194d429b54f38faf6fe/.bazelversion#L1

ThomasHoffmann77 replied:
@boegel Jax 0.6.0 is using Bazel 7.4.1.  Since AF3 is not merged yet, it might be worth to drop jax 0.4.34 and use 0.6.0 instead.

VRehnberg commented:
This patch got rid of the libdevice error for me (for some reason couldn't find Thomas' repo listed to do a PR against ([commit](https://github.com/VRehnberg/easybuild-easyconfigs/commit/9116054450d84f7dd170ccec8e2112fb0ce4d48f))):
```
Find cuda path with EBROOTCUDA. See
https://github.com/easybuilders/easybuild-easyconfigs/pull/21924#issuecomment-2723932545

Author: Viktor Rehnberg (Chalmers University of Technology)

--- /apps/Test/software/jax/0.4.34-gfbf-2024a-CUDA-12.6.0/lib/python3.12/site-packages/jax/_src/lib/__init__.py 2025-02-20 10:12:23.704399435 +0000
+++ -   2025-03-14 08:22:49.679847748 +0000
@@ -18,6 +18,7 @@
 from __future__ import annotations

 import gc
+import os
 import pathlib
 import re
 from typing import Any
@@ -132,7 +133,7 @@
   # If the pip package nvidia-cuda-nvcc-cu11 is installed, it should have
   # both of the things XLA looks for in the cuda path, namely bin/ptxas and
   # nvvm/libdevice/libdevice.10.bc
-  path = _jaxlib_path.parent / "nvidia" / "cuda_nvcc"
+  path = pathlib.Path(os.getenv("EBROOTCUDA", _jaxlib_path.parent / "nvidia" / "cuda_nvcc"))
   if path.is_dir():
     return str(path)
   return None
```

VRehnberg commented:
> is it sufficient to patch the python code only? I have some other patch, which I did not upload yet. It modifies `third_party/tsl/tsl/platform/default/cuda_root_path.cc` allowing to either set CUDA_HOME or XLA_FLAGS=xla_gpu_cuda_data_dir.

Sorry, I don't understand what you're asking. Root issue  is that `path = _jaxlib_path.parent / "nvidia" / "cuda_nvcc"` is not a path that exists as is. Either fix that path or use something else than that function to find CUDA stuff. Those are the two classes of solutions I can imagine. If you've found something else that works, go for it.

ThomasHoffmann77 commented:
> > is it sufficient to patch the python code only? I have some other patch, which I did not upload yet. It modifies `third_party/tsl/tsl/platform/default/cuda_root_path.cc` allowing to either set CUDA_HOME or XLA_FLAGS=xla_gpu_cuda_data_dir.
> 
> Sorry, I don't understand what you're asking. Root issue is that `path = _jaxlib_path.parent / "nvidia" / "cuda_nvcc"` is not a path that exists as is. Either fix that path or use something else than that function to find CUDA stuff. Those are the two classes of solutions I can imagine. If you've found something else that works, go for it.

@VRehnberg I switched to 0.4.35 and added a patch to find libdevice.10.bc relative to $CUDA_HOME

boegel commented:
@ThomasHoffmann77 Any updates on this?

ThomasHoffmann77 commented:
> @ThomasHoffmann77 Any updates on this?


@boegel We have this jax 0.4.35 running with AF3 at EMBL. I did not further work on an update to jax 0.6.0 yet. The current PR still downloads lots of Bazel packages at build time. Some more critical review and testing would be beneficial.

Flamefire commented:
Likely. You can use the same patch as I've done for TF: https://github.com/easybuilders/easybuild-easyconfigs/pull/22921/files#diff-0c447a6b5b271f5000a9d56d61038dba7d149db0435eaf83bc91ead482a47c5f

pavelToman commented:
I just create PR for jax-0.6.2 with CUDA-12.6.0:
- https://github.com/easybuilders/easybuild-easyconfigs/pull/24141
